### DISCUSSION TITLE
Howto pass Program change number unchanged to target?

### DISCUSSION THREAD

**Discussion by ironirc** [helgoboss/helgobox#1297] [Q&A] [answered]
![Scherm­afbeelding 2024-11-01 om 14 58 05](https://github.com/user-attachments/assets/e6c852eb-7440-427c-b520-70b8c6de37ff)

Problem:
The target Max is dynamic - it depends on how many presets you have on the target Fx (Or Container in my case)

I see some possible solutions:
- Add a "No glue" option 
- Option to let "Glue - source max" follow the "Target max"
- Make target max value (an all others) available in Control transformation script

Or I'm missing the obvious solution, which would be the best outcome.

**Comment by helgoboss** [comment, marked as answer]
## Answer

This is the wrong target. "FX: Browse presets" is intended for browsing, with an encoder or with -/+ buttons. Recalling concrete presets gives you exactly the problem that you describe, because absolute control in ReaLearn is always **scaled**, which means that source and target values are proportionally scaled to each other, with the target's value range mapped to the source’s percentage. 

Workaround:

The recommended way to do this is to add one mapping per program change message, using the source [Specific program change](https://docs.helgoboss.org/realearn/sources/midi/specific-program-change.html), combined with the target "FX: Load snapshot".

The "one mapping per program change" strategy is a lot more powerful because the target for each program change message can be configured in very much detail. E.g. if we had a "FX preset: Select" target, you could identify the preset by name instead of by its position. If adding so many mappings is too much clicking for you, you can quite easily generate all those mappings using Lua (similar to [this](https://github.com/helgoboss/helgobox/blob/master/resources/main-presets/notes-to-clip-engine.lua)).

The target "FX: Load snapshot" is usually more desirable if you want to jump to specific presets because it's independent of the FX preset order (changing that order will mess everything up).

That said, if you need a "FX preset: Select" target, feel free to request it (new discussion with type "Idea").

## Background

In the past, I had the idea of introducing a **direct** (non-scaled) mode as well. This corresponds a bit to your "No glue" option, but it still gives the Glue some power. In fact, I made quite some progress as part of #184.

However, I never unlocked that "direct" mode because I thought it adds unnecessary complexity (yet another switch that completely changes the semantics of the Glue section). This increased complexity didn't feel justified because "direct" control is not very common. After all, in all those years, almost nobody requested this behavior. And most importantly, there is above mentioned workaround, which is even more powerful.

**Reply by ironirc** [reply]
Thanks for your elaborate reply Benjamin! 

I've been experimenting with it, and the snapshots work, and they're fast! 

There seems to be a problem though with snap-shotting containers. 
Only the last effect in the container seems to be included in a snapshot.
Or am I doing something wrong?

Another consideration is the size of the snapshots, all being persisted into ~/Library/Application Support/REAPER/presets/vst-helgobox-arm64.ini

I've got a plentitude of snapshots of +/-300Kb.  Is it safe to have an entire live gig setup stored this way?
The ini file is now 6Mb

**Reply by helgoboss** [reply]
It's possible that the target has issues with containers, never tried that. Could you report this as a bug (issues section)?

About the size ... I guess it's a thing that needs to be tried out. 6 MB is still fairly okay, considering how big some RPP files can become. If it gets in the range of several 100 MB, I'd start to worry. If you plan to save hundreds of presets that way, it's maybe not the best solution. I might build another solution for that in the future (something similar to ReaLearn's built-in Pot Browser, not sure if you have discovered that).

**Reply by ironirc** [reply]
I created the new issue: https://github.com/helgoboss/helgobox/issues/1307

I just had a look at POT Browser. This is very cool indeed!!! Looking forward to the new solution. ;-)
